### DISCUSSION TITLE
sp simulation using touchstone file

### DISCUSSION THREAD

**Discussion by caglarozdag** [ra3xdh/qucs_s#1067] [Q&A]
This might be a Xyce issue and not specifically a Qucs-S issue, if that is the case apologies in advance.

Long story short, I need a flow to EM simulate multi port designs and do sparameter and transient simulations on the multi-port sp (touchstone) file. I figured out that OpenEMS and Xyce is my best choice. I am using Qucs-S for my simulator GUI.

I confirmed that I successfully integrated Xyce and Qucs-S because I was able to simulate a simple inductor S11 this way.

But my main task is to run sp simulation on a touchstone file (example below is for a 2 port touchstone file, but eventually I need to be able to run multi port simulations).

The following netlist and attached schematic screenshot results in an empty results database. As you can see in the screenshot S11 is all 1s.

```
* Qucs 24.3.2  /home/caglarozdag/qucs_s_tests/2portspsim_xyce.sch
P1 _net0 0 port=1  z0=50  AC 0.632456  SIN 0 0.632456 1MEG
P2 _net1 0 port=2  z0=50  AC 0.632456  SIN 0 0.632456 1MEG
YLIN YLIN_X1 _net0 0 _net1 0 YLIN_X1_model
.MODEL YLIN_X1_model LIN TSTONEFILE=/home/caglarozdag/OpenEMStest/test1/data3/run_2port_simpletest2_2port_network.s2p
.AC LIN 201 1MEG 30G
.LIN format=touchstone sparcalc=1
.PRINT ac format=std file=spice4qucs_sparam.prn sdb(1,1) s(1,1) sp(1,1) y(1,1) z(1,1) sdb(1,2) s(1,2) sp(1,2) y(1,2) z(1,2) sdb(2,1) s(2,1) sp(2,1) y(2,1) z(2,1) sdb(2,2) s(2,2) sp(2,2) y(2,2) z(2,2)
.END
```

![sp_schematic](https://github.com/user-attachments/assets/778e1bfd-c8c3-42bd-9754-54113cfa9121)

Can someone help me debug my issue? Alternatively, I could directly run a Xyce netlist from command line, but no luck there either. I would highly prefer to use the Qucs-S GUI. Any help is appreciated.

**Comment by ra3xdh** [comment]
XYCE doens't support the touchsotne files with S-parameter simulation. There is no workaround to simulate it with Xyce. The S-parameter device is mapped by Qucs-S to Xyce YLIN device. The YLIN works only with HB analysis and gives zero output otherwise. Please read the section 2.3.24 of the Xyce Reference guide.

**Reply by caglarozdag** [reply]
I see... Thanks. Xyce was my only hope... Is there any way that you know (within or outside of Qucs-S) that can simulate multi port touchstone files (not just 2 port)? Apparently Ngspice cant.

**Reply by ra3xdh** [reply]
Ngspice has no direct support for touchstone file, but Qucs-S supports automatic conversion of these files to XFER XSPICE sources. See #233 for more info. Unfortunately, this feature still be not documented.

**Reply by caglarozdag** [reply]
I have used ngspice (as you said, it called xfer) for 2 port touchstone file simulations with success. But I absolutely need multi port capability (>2 ports support. Do you know this method that calls ngspice xfer is strictly limited to 2 ports, or maybe I am doing something else wrong (screenshot attached of netlist error).
![sp_ngspice_3port_problem](https://github.com/user-attachments/assets/4dfcfe66-ae86-4313-bc5b-4dc8505ecc9d)

**Comment by tomhajjar** [comment]
ngspice supports more than 2 ports. I have a few examples using more than 2.

The s-parameter files are converted to Spice compatible sources. That is now done automatically by Qucs-S but there are a few programs that do the conversion. I prefer the one done for Windows called S2spice by Dan Dckey. @transmitterdan S2spice support many ports and a large number of frequency points.

https://github.com/transmitterdan/s2spice

![2024-11-17_140907](https://github.com/user-attachments/assets/83f81622-1abb-42d7-8933-5bf4ec8f6127)

[Transmission_Line_prj.zip](https://github.com/user-attachments/files/17792034/Transmission_Line_prj.zip)

**Comment by tomhajjar** [comment]
Here is another. Mini Circuits Balun, 3-ports. Circuit was done before Qucs-S could process S-parameter files. ADTL2_18.s3p was converted by S2spice to ADTL2_18.lib.

Originally on LTspice Forum.
https://groups.io/g/LTspice/topic/92954449#msg139714

![2024-11-17_212207](https://github.com/user-attachments/assets/7767134a-5058-468b-af5b-1f93bbc79453)

[S-parameter_s3p.zip](https://github.com/user-attachments/files/17793603/S-parameter_s3p.zip)

**Comment by tomhajjar** [comment]
Finally.... Mini Circuits 90deg Hybrid s4p file

![2024-11-17_221649](https://github.com/user-attachments/assets/8a2604e5-9326-4ef0-be93-c574de29c022)

[ADQ_22_Hybrid.zip](https://github.com/user-attachments/files/17794177/ADQ_22_Hybrid.zip)
